api/src/main/java/org/openmrs/module/labtrackingapp/api/impl/LabTrackingAppServiceImpl.java

```diff
@@ -36,6 +39,80 @@ public List<Encounter> getSpecimenDetailsEncountersByOrderNumbers(String[] order
         return dao.getSpecimenDetailsEncountersByOrderNumbers(orderNumbers);
     }
 
+    public List<Encounter> getSpecimenDetailsEncountersByDate(long startDate, long endDate, String patientUuid, String patientName, int status, boolean suspectedCancer, boolean urgentReview, int maxResults) {
+
+        List<Encounter> encounters = null;
+        List<OrderType> orderTypes = new ArrayList<OrderType>();
+        orderTypes.add(Context.getOrderService().getOrderTypeByUuid(LabTrackingConstants.LAB_TRACKING_PATHOLOGY_ORDER_TYPE_UUID));
+
+        List<Order> orders = dao.getOrdersByDate(startDate, endDate, patientUuid, patientName, status,  orderTypes, maxResults);
+        if ( orders != null && orders.size() > 0 ) {
+            encounters = new ArrayList<Encounter>();
+            for (Order order : orders) {
+                Encounter encounter = order.getEncounter();
+                boolean includeEncounter = true;
+
+                if ( StringUtils.isBlank(patientUuid)
+                        && StringUtils.isBlank(patientName)
+                        && (status == LabTrackingConstants.LabTrackingOrderStatus.ALL.getId() || status == LabTrackingConstants.LabTrackingOrderStatus.CANCELED.getId() )
+                        && !suspectedCancer && !urgentReview) {
+                    // if the status is CANCELED the encounter was already filtered by the Order DAO method
+                    encounters.add(encounter);
+                    continue;
+                }
+                if (LabTrackingConstants.LabTrackingOrderStatus.REQUESTED.getId() == status) {
+                    Obs obs = getFirstObsByConceptUuid(encounter, LabTrackingConstants.LAB_TRACKING_SPECIMEN_ENCOUNTER_PROCESSED_DATE_UUID);
+                    if ( obs != null && obs.getValueDatetime() != null) {
+                        includeEncounter = false;
+                    }
+                } else if (LabTrackingConstants.LabTrackingOrderStatus.PROCESSED.getId() == status) {
+                    Obs processedDate = getFirstObsByConceptUuid(encounter, LabTrackingConstants.LAB_TRACKING_SPECIMEN_ENCOUNTER_PROCESSED_DATE_UUID);
+                    Obs resultDate = getFirstObsByConceptUuid(encounter, LabTrackingConstants.LAB_TRACKING_SPECIMEN_ENCOUNTER_RESULTS_DATE_UUID);
+                    if ( processedDate == null || processedDate.getValueDatetime() == null || resultDate != null ) {
```

Comment:
I *definitely* might be wrong about this, but shouldn't conditional before the resultDate be an AND...

ie, isn't the logic, "if this encounter does not have a processed dated, don't show it,*unless* it has a result date, in which case we assume it has been processed, even if it doesn't have a processed date?"